hw/opentitan/ot_hmac.c

```diff
@@ -640,17 +914,27 @@ static void ot_hmac_regs_write(void *opaque, hwaddr addr, uint64_t value,
 
             ibex_irq_set(&s->clkmgr, true);
 
+            /* Hold the previous digest size until the HMAC is started with the
+            new digest size configured */
+            s->ctx->digest_size_started = ot_hmac_get_digest_size(s->regs->cfg);
+
             ot_hmac_sha_init(s, true);
 
             /* HMAC mode, process input padding */
             if (s->regs->cfg & R_CFG_HMAC_EN_MASK) {
-                uint64_t ipad[8u];
+                size_t key_length_b = ot_hmac_get_key_bytes(s);
+                size_t block_size_b = ot_hmac_get_block_size_bytes(s);
+                /* If the key is smaller than the block size, we must pad it to
+                the right with 0s. */
+                size_t pad_length_b = MAX(key_length_b, block_size_b);
+                size_t pad_length_w = pad_length_b / sizeof(uint32_t);
+                uint32_t ipad[OT_HMAC_MAX_KEY_LENGTH / sizeof(uint32_t)];
```

Comment:
Can we have length/size which are not a multiple of 8 bytes? If not, we could use `uint64_t` values here. Just checking :)

Reply:
From my understanding, all lengths/sizes are multiples of 16 bytes, so this can definitely be changed - I mostly used `uint32_t` to match the type width of `s->regs->key`, as it made the logic easier to follow in my head. Is there a reason that you would prefer `uint64_t` here?

Reply:
This is the native size for most hosts nowadays (x86_64, aarch64). Micro optimization but the less time/loops here the faster the results and/or the more time available for all the other IPs in the I/O thread. HMAC implementation is a known bottleneck at the moment on some workloads (such as verifying large FW/data chunks)

Reply:
That makes sense, thanks for the detailed explanation. I'll change that back to using `uint64_t`.

Reply:
Thanks.